stix_shifter_utils/utils/param_validator.py

```diff
@@ -0,0 +1,149 @@
+
+def del_dot_path(params, path):
+    if '.' in path:
+        pp = path.split('.')
+        if pp[0] in params:
+            del_dot_path(params[pp[0]], '.'.join(pp[1:]))
+            if len(params[pp[0]]) == 0:
+                del params[pp[0]]
+    else:
+        del params[path]
+
+def param_validator(module, input_configs):
+
+    config_json_path = f'stix_shifter_modules/{module}/configuration/{module}_config.json'
+    
+    try:
+        with open(config_json_path) as mapping_file:
+            expected_configs = json.load(mapping_file)
+    except Exception as ex:
+        raise(ex)
+    
+    validated_params = {}
+    errors = []
+    copy_valid_configs(input_configs, expected_configs, validated_params, errors)
+
+    error_obj = {}
+    if errors:
+        error_obj['missing_params'] = errors
+    if input_configs:
+        if isinstance(input_configs, dict):
+            error_obj['unexpected_params'] = get_inner_keys(input_configs)
+
+    if error_obj:
+        raise ValueError(error_obj)
+
+    return validated_params
+
+
+def copy_valid_configs(input_configs, expected_configs, validated_params, errors=[], current_path=''):
+    if isinstance(expected_configs, dict):
+        for key, value in expected_configs.items():
+            key_path = current_path
+            if key_path:
+                key_path = key_path + '.'
+            key_path = key_path + key
+            if key in input_configs:
+                if is_leaf(expected_configs[key]):
+                    if 'min' in expected_configs[key]:
+                        if not check_min(input_configs[key], expected_configs[key]['min']):
+                            raise ValueError('\"{}: {}\" value must be more than {}'.format(key, str(input_configs[key]), str(expected_configs[key]['min'])))
+                    elif 'max' in expected_configs[key]:
+                        if not check_max(input_configs[key], expected_configs[key]['max']):
+                            raise ValueError('\"{}: {}\" value must be less than {}'.format(key, str(input_configs[key]), str(expected_configs[key]['max'])))
+                    elif 'regex' in expected_configs[key]:
+                        if not check_regex(input_configs[key], expected_configs[key]['regex']):
+                            raise ValueError('Invalid {} value \"{}\" specified'.format(key, str(input_configs[key])))
+
+                    validated_params[key] = input_configs[key]
+                    del input_configs[key]
+                else:
+                    if key not in validated_params:
+                        validated_params[key] = dict()
+                    copy_valid_configs(input_configs[key], expected_configs[key], validated_params[key], errors, key_path)
+                    if not input_configs[key]:
+                        del input_configs[key]
+            else:
+                if not 'optional' in expected_configs[key] and not 'default' in expected_configs[key]:
```

Comment:
expect it to add an error even if there is optional param in options but with 'False' value.
107-111 can be replaced with something like that:
if 'default' in expected_configs[key]:
    validated_params[key] = value['default']
elif not ('optional' in expected_configs[key] or expected_configs[key]['optional']):
    errors.append(key_path)

Reply:
your conditions validates if there is 'optional' present in the configuration it does not check the value itself.

Reply:
for our use cases, we only specify `optional` when the value is True. I think we had a discussion to remove `"optional": true`. You suggestion may not work. I need to change the conditional check a little bit